### PR TITLE
[WIP] Refactor floats

### PR DESCRIPTION
**Pull request by clason**
The current preview implementation in `output.vim` has gotten a bit unwieldy in trying to support vim popups, neovim floats, and preview buffers, making it difficult to extend or adapt to changes in the low-level APIs (in particular for the upcoming neovim 0.4.0).

This PR proposes a refactor to completely separate the backend-specific details from the functions handling the server response. The idea is to have

1. a **high-level API** for showing and updating "tooltips" (here used as a generic name to distinguish from, e.g., neovim floats -- all names up for discussion) and
2. an **intermediate-level API** for creating, updating, positioning, and closing tooltips, implemented separately for each backend using the respective low-level API.

The current goal of this PR is to solicit discussion on the design of these APIs, documented in `doc/tooltip-api.txt`. (The code is meant as a stub to show the proposed dispatch and to have somewhere for sketching out implementation ideas.) The most central questions are:

1. The high-level API proposal has two separate functions for "at-cursor" (hover, peek, signature) and "at-pum" (documentation) tooltips. Is this enough? Or should these be combined?
2. Which options should these functions support (border, relative position, width, autocommands,...)?

### CODE REVIEW

File: doc/tooltip-api.txt [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+In addition, there are common utility functions:
+
+get_size_info({lines})
+     
+     Get width and height needed to show the line-wrapped contents of a list
+     of strings.
+
+     Returns a tuple of |[lines,width]| 
+
+     TODO: in characters or bytes? 0- or 1-indexed?
```
Comment (thomasfaingnaert) [inline doc/tooltip-api.txt]: 0 and 1 indexing doesn't make much sense for sizes ;)

Comment (clason) [reply doc/tooltip-api.txt]: True :) I was thinking here of utility functions in general, such as ones that compute position (for moving tooltips away from the buffer boundary).

**Comment by thomasfaingnaert** [comment]
Commenting for future reference: https://github.com/vim/vim/commit/dca7abe79cc4f0933473c3e4bcc75b46cc2c48fd may be useful later.

**Comment by hrsh7th** [comment]
Hi there, I have floatwin/popup-win compatible codes.
The codes is similar to the proposed apis (but bit different).

If @clason don't mind, I will send PR with implementation (not spec).

**Comment by clason** [comment]
@hrsh7th That would be very much appreciated; I have to admit I put this on a back burner due to real-life issues and the fact that neovim looks to be introducing new (and more convenient) APIs in the current release cycle (0.5), which looked worth waiting for.

The main point of this refactor is to separate the API used by vim-lsp's callbacks from the backends, and as long as your implementation does the same (in whatever way), it's fine.

**Comment by hrsh7th** [comment]
Thanks for your instructuon.

I started to switching to new implementation.
But it's bit difficult (syntax highlight etc...), So it may take sometime.

**Comment by hrsh7th** [comment]
I've sent the PR. Please check and feedback #567.

**Comment by prabirshrestha** [comment]
let us continue the discussion at https://github.com/vim-jp/vital.vim/issues/747 and https://github.com/vim-jp/vital.vim/pull/748.

**Pull request closed without merging**